src/nbl/asset/interchange/CPLYMeshWriter.cpp

```diff
@@ -78,44 +78,81 @@ bool CPLYMeshWriter::writeAsset(io::IWriteFile* _file, const SAssetWriteParams&
+        bool doesItHaveIndexBuffer = inputMeshBuffer->getIndexBufferBinding().buffer.get(), isItNotTriangleListsPrimitive = inputMeshBuffer->getPipeline()->getPrimitiveAssemblyParams().primitiveType != asset::EPT_TRIANGLE_LIST;
+        
+        if (doesItHaveIndexBuffer && isItNotTriangleListsPrimitive)
+        {
+            auto cpuConvertedMeshBuffer = core::smart_refctd_ptr_static_cast<asset::ICPUMeshBuffer>(inputMeshBuffer->clone());
+            IMeshManipulator::homogenizePrimitiveTypeAndIndices(&cpuConvertedMeshBuffer, &cpuConvertedMeshBuffer + 1, asset::EPT_TRIANGLE_LIST, asset::EIT_32BIT);
+            return cpuConvertedMeshBuffer;
+        }
+        else
+            return nullptr;
+    };
+
+    const auto cpuConvertedMeshBufferWithIndexBuffer = getConvertedCpuMeshBufferWithIndexBuffer();
+    const asset::ICPUMeshBuffer* rawCopyMeshBuffer = cpuConvertedMeshBufferWithIndexBuffer.get() ? cpuConvertedMeshBufferWithIndexBuffer.get() : *meshbuffers.begin();
+    const bool doesItUseIndexBufferBinding = (rawCopyMeshBuffer->getIndexBufferBinding().buffer.get() && rawCopyMeshBuffer->getIndexType() != asset::EIT_UNKNOWN);
+
+    uint32_t faceCount = {}; 
+    size_t vertexCount = {};
+
+    void* indices = nullptr;
+    {
+        auto indexCount = rawCopyMeshBuffer->getIndexCount();
+
+        indices = _NBL_ALIGNED_MALLOC(indexCount * sizeof(uint32_t), _NBL_SIMD_ALIGNMENT);
+        memcpy(indices, rawCopyMeshBuffer->getIndices(), indexCount * sizeof(uint32_t));
+        
+        IMeshManipulator::getPolyCount(faceCount, rawCopyMeshBuffer);
+        vertexCount = IMeshManipulator::upperBoundVertexID(rawCopyMeshBuffer);
+    }
+
 	// write PLY header
     std::string header = "ply\n";
     header += (flags & asset::EWF_BINARY) ? "format binary_little_endian 1.0" : "format ascii 1.0";
 	header += "\ncomment IrrlichtBAW ";
 	header +=  NABLA_SDK_VERSION;
 
-    auto meshBuffer = *meshbuffers.begin();
-	// get vertex and triangle counts
-	size_t vtxCount = IMeshManipulator::upperBoundVertexID(meshBuffer);
-    size_t faceCount = meshBuffer->getIndexCount() / 3;
-
 	// vertex definition
 	header += "\nelement vertex ";
-	header += std::to_string(vtxCount) + '\n';
+	header += std::to_string(vertexCount) + '\n';
 
     bool vaidToWrite[4]{ 0, 0, 0, 0 };
-    auto mbPipeline = meshBuffer->getPipeline();
 
-    if (meshBuffer->getAttribBoundBuffer(0).buffer)
+    if (rawCopyMeshBuffer->getAttribBoundBuffer(0).buffer)
     {
-        const asset::E_FORMAT t = meshBuffer->getAttribFormat(0);
+        const asset::E_FORMAT t = rawCopyMeshBuffer->getAttribFormat(0);
         std::string typeStr = getTypeString(t);
         vaidToWrite[0] = true;
         header +=
             "property " + typeStr + " x\n" +
             "property " + typeStr + " y\n" +
             "property " + typeStr + " z\n";
     }
-    if (meshBuffer->getAttribBoundBuffer(1).buffer)
+    if (rawCopyMeshBuffer->getAttribBoundBuffer(1).buffer)
```

Comment:
wtf why are integer literals being used here?

ICPUMeshBuffer reports the attribute IDs used for position and normals, so use these.

For UV and Color you can assume 2 and 1 but make some `constexpr` with a proper name

Reply:
todonigga

Reply:
corrected in following commit